src/y0/algorithm/estimation.py

```diff
@@ -0,0 +1,334 @@
+    treatment_value,
+    outcome: Variable,
+) -> float:
+    """Estimate the counterfactual mean E[Y(t)] with the Primal IPW estimator."""
+    beta_primal = get_beta_primal(
+        data=data,
+        graph=graph,
+        treatment=treatment,
+        treatment_value=treatment_value,
+        outcome=outcome,
+    )
+    return np.mean(beta_primal).item()
+
+
+def fit_binary_model(data, formula, weights=None) -> GLM:
+    """Fit a binary general linear model."""
+    return GLM.from_formula(
+        formula,
+        data=data,
+        family=Binomial(),
+        freq_weights=weights,
+    ).fit()
+
+
+def fit_continuous_glm(data, formula, weights=None) -> GLM:
+    """Fit a continuous general linear model."""
+    return GLM.from_formula(
+        formula,
+        data=data,
+        family=Gaussian(),
+        freq_weights=weights,
+    ).fit()
+
+
+def get_beta_primal(
+    *,
+    data: pd.DataFrame,
+    graph: NxMixedGraph,
+    treatment: Variable,
+    outcome: Variable,
+    treatment_value,
+    state_space_map: Dict[Variable, str],
+) -> np.array:
+    """Return the beta primal value for each row in the data.
+
+    This code was adapted from :mod:`ananke` ananke code at:
+    https://gitlab.com/causal/ananke/-/blob/dev/ananke/estimation/counterfactual_mean.py?ref_type=heads#L408-513
+
+    :param data: Given data
+    :param graph: A NxMixedGraph
+    :param treatment: Given treatment
+    :param outcome: Given outcome
+    :param treatment_value: Given treatment value
+    :param state_space_map: Contains the state space of a variable (binary/continuous)
+    :returns: np.array
+    """
+    # extract the outcome
+    y = data[outcome.name]
+
+    # c := pre-treatment vars and l := post-treatment vars in district of treatment
+    c = graph.pre(treatment)
+    post = set(graph.nodes()).difference(c)
+    l = post.intersection(graph.get_district(treatment))
+
+    # create copies of the data with treatment assignments t=0 and t=1
+    data_t1 = data.copy()
+    data_t1[treatment.name] = 1
+    data_t0 = data.copy()
+    data_t0[treatment.name] = 0
+
+    indices = data[treatment.name] == treatment_value
+
+    # prob: stores \prod_{li in l} p(li | mp(li))
+    # prob_t1: stores \prod_{li in l} p(li | mp(li)) at t=1
+    # prob_t0: stores \prod_{li in l} p(li | mp(li)) at t=0
+
+    mp_t = graph.get_markov_pillow([treatment])
+    indices_t0 = data.index[data[treatment.name] == 0]
+
+    if len(mp_t) != 0:
+        formula = treatment.name + " ~ " + "+".join([variable.name for variable in mp_t])
+        model = fit_binary_model(data, formula)
+        prob = model.predict(data)
+        prob[indices_t0] = 1 - prob[indices_t0]
+        prob_t1 = model.predict(data)
+        prob_t0 = 1 - prob_t1
+    else:
+        prob = np.ones(len(data)) * np.mean(data[treatment.name])
+        prob[indices_t0] = 1 - prob[indices_t0]
+        prob_t1 = np.ones(len(data)) * np.mean(data[treatment.name])
+        prob_t0 = 1 - prob_t1
```

Comment:
can you refactor the parts of both of these conditions that are the same out of the condition?

`prob[indices_t0] = 1 - prob[indices_t0]` and `prob_t0 = 1 - prob_t1` appear to be the same

Reply:
Sure